src/main/java/com/google/sps/servlets/RegistrationServlet.java

```diff
@@ -26,23 +34,61 @@
-    String role = request.getParameter("role");
-    String name = request.getParameter("name");
-    String email = request.getParameter("email");
+    String role = getParameter(request, "role").orElse(null);
+    String firstName = getParameter(request, "first-name")
+            .orElseThrow(() -> new IllegalArgumentException("Must fill out first name"));
+    String lastName = getParameter(request, "last-name")
+            .orElseThrow(() -> new IllegalArgumentException("Must fill out last name"));
+    String fullName = firstName + " " + lastName;
+    
+    String email = userService.getCurrentUser().getEmail();
+    String userId = userService.getCurrentUser().getUserId();
+    
     // Make list of selected topics, remove unchecked topics
-    List<String> topics = Arrays.asList(request.getParameter("math"), request.getParameter("english"), request.getParameter("other"));
-    topics = topics
-            .stream()
-            .filter(t -> t!= null)
-            .collect(Collectors.toList());
+    List<Optional<String>> topics = new ArrayList<Optional<String>>();
+    topics.add(getParameter(request, "math"));
+    topics.add(getParameter(request, "english"));
+    topics.add(getParameter(request, "other"));
+    List<String> topicsToStr = topics
+                                .stream()
+                                .filter(t -> t.isPresent())
+                                .map(t -> t.get())
+                                .collect(Collectors.toList());
 
-    response.setContentType("text/html;");
-    response.getWriter().println("<p>Student or tutor? " + role + "</p>");
-    response.getWriter().println("<p>Name: " + name + "</p>");
-    response.getWriter().println("<p>Email: " + email + "</p>");
-    response.getWriter().println("<p>Subject(s): " + topics + "</p>");
+    // Make entity for user with all registration info
+    Entity userEntity = new Entity("User");
+    createEntityAndPutInDatastore(datastore, userEntity, role, fullName, email, userId, topicsToStr);
+
+    response.sendRedirect("/scheduling.html");
+  }
+
+ /**
+  * Creates a user entity and puts it in datastore, used for testing
+  */
+  public void createEntityAndPutInDatastore(DatastoreService ds, Entity entity, String role, String name, String email, String userId, List<String> topics) {
+    entity.setProperty("role", role);
+    entity.setProperty("name", name);
+    entity.setProperty("email", email);
+    entity.setProperty("userId", userId);
+    entity.setProperty("topics", topics);
+    ds.put(entity);
   }
+  
+  /**
+   * @return the Optional of the request parameter
+   */
+   private Optional<String> getParameter(HttpServletRequest request, String name) {
+       String value = request.getParameter(name);
+       //return empty Optional if string is null or empty
+       if (value == null || value.isEmpty()) {
+           return Optional.empty();
+       }
+       return Optional.<String>ofNullable(value);
+   }
```

Comment:
Instead of having a separate function for get parameter, you could also just use "Optional.ofNullable(request.getParameter(...)).orElse(...);"

Reply:
Good idea, that makes the code a lot more concise